### ISSUE TITLE
fi_info doesn't show the verbs rdm provider

### ISSUE THREAD

**Issue by JoZie** [ofiwg/libfabric#2556]
I'm developing an application that uses a RDM endpoint. Now I want to test it on our HPC-System.
Therefor I wanted to check the existing providers with the `fi_info` tool. But the output showed only the verbs `FI_EP_MSG` provider.

I did a little debugging [(output)](https://gist.github.com/JoZie/2928142eb0789f5656bcf93cb99552d6) and discovered that there is a missmatch of ip addresses in `fi_ibv_retain_info` (lines 23 and 29 of the output). If I'm providing the correct ip along with the call `fi_info -n 172.24.0.16` everything seems to be fine [(output)](https://gist.github.com/JoZie/c7a2284a58257ab0fdf3ada174945b61).

On our system a call to `ifconfig` shows two devices `ib0 (172.24.0.16)` and the virtual `ib0:0 (172.24.16.16)`  with the same hardware address.   

The virtual device somehow prevents libfabric from selecting the verbs provider with the real device (`ib0`).
The same applies to my application. I can only use it if i'm providing `FI_VERBS_IFACE="ib0:0"` in the environment.

**Comment by a-ilango** [comment]
@evgeny-leksikov can you please take a look?

**Comment by evgeny-leksikov** [comment]
@JoZie Thanks for reporting. Currently, the verbs/rdm does not have any intelligent recognition of available interfaces. It takes first match by "ib" name and works with that. Using FI_VERBS_IFACE parameter is right way for definition of desired interface in current version.
Could you share initialization config files from your system (for RHEL, they are placed in /etc/sysconfig/network-scripts/ifcfg-ib*) It would be helpful to reproduce and investigate your issue for future improvements.

**Comment by JoZie** [comment]
@evgeny-leksikov 
Thanks for your answer. Even though I uploaded the two [config files](https://gist.github.com/JoZie/e4c6e2532426c24a202d9e2b6651cf36) from our login nodes, the rest of the system is configured similary.
I tried to use the FI_VERBS_IFACE parameter with the correct interface name (`ib0`) but this doesn't work, there is still no verbs-rdm provider available. The debug output looks basically the same as [the first one](https://gist.github.com/JoZie/2928142eb0789f5656bcf93cb99552d6) I posted. 
I assume that there is somehow a substring matching involved during the resolving of the interface name?

**Comment by evgeny-leksikov** [comment]
@JoZie Thanks! Yes, you are right, FI_VERBS_IFACE works as a prefix and, in your case, both start from ib0 but ib0:0 is matching first.
@hoopoepg please prepare a fix for strong matching when it's available in the system.

**Comment by hoopoepg** [comment]
@JoZie could you try this PR: https://github.com/ofiwg/libfabric/pull/2570
we bit updated initialization scheme which allows exact matching of specified interface

thank you

**Comment by JoZie** [comment]
@hoopoepg Thanks for your help!
I tried your PR and it brought back the verbs RDM provider but unfortunately with the wrong device.
The value of the FI_VERBS_IFACE variable had no influence anymore because the verbs provider always returned the information from the last interface found. On our system the order is `ib0` `ib0:0`.

To fix the issue myself I had prevent assigning the wrong source address to the `fi_info` struct.  Therefore I used your PR as foundation and modified  the `fi_ibv_getifaddrs` function to also use the value of FI_VERBS_IFACE for selecting the address . You can find the commit JoZie@91650d7290e336187c2b818fa0f1e2be88afda00 or should I open a PR myself?

**Comment by hoopoepg** [comment]
@JoZie 
it seems I uploaded non-complete fix :)
thank you for investigation. I added your patch into my version of fix, could you try it https://github.com/ofiwg/libfabric/pull/2570 ?

unfortunately we have no system where your configuration can be reproduced.
let me know about results

thank you

**Comment by JoZie** [comment]
@hoopoepg 

TLDR; 
There is a bug with a missing null-terminator at line 1110.
It still doesn't work correctly because of effects from `fi_ibv_retain_info` and `fi_ibv_getifaddrs`.

---
I just tried your version and it triggers a bug so that the `strncmp` doesn't return a correct "exact match".
The bug is that  the `iface` string might not have a null-terminator. That's because `strlen(iface_tmp)` ignores it and `strncpy` doesn't handle it either. So line 1100 should look like:
```
         strncpy(iface, iface_tmp, iface_len+1);
```
This relies on the fact that `getenv` returns a null-terminated string.

With this bug fixed the verbs RDM provider doesn't show up anymore, because the current state of your PR touches not all parts of the decision. My investigation earlier today showed that the second  dependency for returning the RDM provider for `ib0` is not fulfilled. Everything works fine with `ib0:0`. Let me explain.

The source code path used on our machine is roughly the following:
```
1230:        ret = fi_ibv_init_info();
1234:        ret = fi_ibv_create_ep(node, service, flags, hints, &rai, &id);
1242:                ret = fi_ibv_get_matching_info(NULL, hints, rai, info);
1245:                        ret = fi_ibv_getifaddrs(service, flags, *info);
1255:                ret = fi_ibv_rdm_remove_nonaddr_info(info);
1193:        ret = fi_ibv_rdm_find_sysaddrs(&iface_addr, &lo_addr);
```
The value for `iface_addr` now matches the address of the interface specified in FI_VERBS_IFACE. The execution goes on with:
```
1200:        while (fi) {
1201:                if (fi_ibv_retain_info(fi, &iface_addr, &lo_addr)) {
1220:        return *info ? 0 : -FI_ENODEV;
```
The while loop at line 1200 iterates over all `fi_info` structs and retains/removes structs based on the comparison of `iface_addr` and `fi->src_addr`. The problem is that the hints of  my code (get me the verbs RDM provider) lead to the situation that there is only one `fi_info` struct with the wrong value for `src_addr`. Because of that the `fi_info` struct will be removed and `fi_getinfo` will return that no provider is available. Of course the same applies to fi_info.

The cause for that is, that  `fi_ibv_getifaddrs` at line 1245 sets the `src_addr` to the value of the last found interface. In my case this is the ip of `ib0:0` not ! the one for the specified interface (`ib0`). Therefore my patch also applies your exact match test to `fi_ibv_getifaddrs`.

I hope I could explain to you why my fix is larger than yours and why I think every part is necessary. Even though you cant reproduce my machine settings. I'm sure there is another solution for the unfortunate effects from `fi_ibv_retain_info` and `fi_ibv_getifaddrs`. :)

thank you very much

**Comment by a-ilango** [comment]
> The cause for that is, that fi_ibv_getifaddrs at line 1245 sets the src_addr to the value of the last found interface. In my case this is the ip of ib0:0 not ! the one for the specified interface (ib0). Therefore my patch also applies your exact match test to fi_ibv_getifaddrs.

This is probably not right behavior on part of fi_ibv_getifaddrs. Perhaps the function should insert an additional fi_info struct corresponding to the virtual interface.

**Comment by hoopoepg** [comment]
there is no way to find matching between IB device and configured IP interface, that is why we have to implement hint-based logic to select proper IP interface

I updated PR https://github.com/ofiwg/libfabric/pull/2570 where issue should be fixed (we reproduced issue on our side and it works at least on our side :) )
